lib/lightning_web/channels/ai_assistant_channel.ex

```diff
@@ -0,0 +1,775 @@
+      Enum.reduce(opts, msg, fn {key, value}, acc ->
+        String.replace(acc, "%{#{key}}", to_string(value))
+      end)
+    end)
+    |> flatten_association_errors()
+  end
+
+  defp flatten_association_errors(errors) do
+    Enum.reduce(errors, %{}, fn {key, value}, acc ->
+      flatten_error_value(key, value, acc)
+    end)
+  end
+
+  @doc false
+  def flatten_error_value(key, list, acc) when is_list(list) do
+    if Enum.any?(list, &is_map/1) do
+      flatten_nested_list_errors(key, list, acc)
+    else
+      Map.put(acc, to_string(key), list)
+    end
+  end
+
+  @doc false
+  def flatten_error_value(key, value, acc) do
+    Map.put(acc, to_string(key), value)
+  end
+
+  defp flatten_nested_list_errors(key, list, acc) do
+    list
+    |> Enum.with_index()
+    |> Enum.reduce(acc, fn {item_errors, index}, inner_acc ->
+      flatten_item_errors(key, item_errors, index, inner_acc)
+    end)
+  end
+
+  defp flatten_item_errors(key, item_errors, index, acc) do
+    Enum.reduce(item_errors, acc, fn {field, messages}, nested_acc ->
+      flattened_key = "#{key}[#{index}].#{field}"
+      Map.put(nested_acc, flattened_key, messages)
+    end)
+  end
+
+  defp get_workflow_for_session(session) do
+    cond do
+      session.meta["unsaved_workflow"] ->
+        nil
+
+      session.workflow_id ->
+        Workflows.get_workflow(session.workflow_id)
+
+      true ->
+        nil
+    end
+  end
+
+  defp get_resource_for_session_type("job_code", session) do
+    job_id =
+      cond do
+        session.job_id -> session.job_id
+        session.meta["unsaved_job"] -> session.meta["unsaved_job"]["id"]
+        true -> nil
+      end
+
+    if job_id do
+      {:ok, job_id}
+    else
+      {:error, "Job not found"}
+    end
+  end
+
+  defp get_resource_for_session_type("workflow_template", session) do
+    case session.project_id do
+      nil -> {:error, "Project not found"}
+      project_id -> {:ok, Projects.get_project(project_id)}
+    end
+  end
+
+  defp format_session(session) do
+    base = %{
+      id: session.id,
+      title: session.title,
+      session_type: session.session_type,
+      message_count: session.message_count || 0,
+      updated_at: session.updated_at
+    }
+
+    case session.session_type do
+      "job_code" ->
+        format_job_code_session(base, session)
+
+      "workflow_template" ->
+        format_workflow_template_session(base, session)
+    end
+  end
+
+  defp format_job_code_session(base, session) do
+    cond do
+      session.meta["unsaved_job"] ->
+        unsaved_job = session.meta["unsaved_job"]
+        workflow = Workflows.get_workflow(unsaved_job["workflow_id"])
+
+        Map.merge(base, %{
+          job_name: unsaved_job["name"],
+          workflow_name: workflow.name,
+          is_unsaved: true
+        })
+
+      session.job_id ->
+        case Jobs.get_job(session.job_id) do
+          {:ok, job} ->
+            workflow = Workflows.get_workflow(job.workflow_id)
+
```

Comment:
Looks like we are introducing and N+1 issue here. When you list say 20 sessions, thats 40+ extra queries. Should we not push the preloading into `list_sessions/3` and make this function purely view/rendering?